File: app/src/main/java/nl/jeroen/dayclock/MainActivity.java
package nl.jeroen.dayclock;

import android.os.Bundle;
import android.os.Handler;
import android.support.v7.app.AppCompatActivity;
import android.widget.TextView;

import java.time.LocalDateTime;
import java.time.temporal.ChronoUnit;
import java.util.Locale;
import java.util.concurrent.TimeUnit;

import nl.jeroen.dayclock.model.TimeDataContainer;

public class MainActivity extends AppCompatActivity {

    private static final long MsInDay = 24 * 60 * 60 * 1000;
    private static final int Delay = 0;
    private static final int Period = 1000;
    private static final String Format = "%02d:%02d:%02d:%03d";

    private TextView tvTime;
    private TextView tvPercent;
    private ClockView clockView;

    private Handler timerHandler = new Handler();
    private Runnable timerRunnable = new Runnable() {
        @Override
        public void run() {
            CalculateYearTime();

            timerHandler.postDelayed(this, Period);
        }
    };

    private static TimeDataContainer getDurationBreakdown(long millis) {
        if (millis < 0) {
            throw new IllegalArgumentException("Duration must be greater than zero!");
        }

        long hours = TimeUnit.MILLISECONDS.toHours(millis);
        millis -= TimeUnit.HOURS.toMillis(hours);
        long minutes = TimeUnit.MILLISECONDS.toMinutes(millis);
        millis -= TimeUnit.MINUTES.toMillis(minutes);
        long seconds = TimeUnit.MILLISECONDS.toSeconds(millis);
        millis -= TimeUnit.SECONDS.toMillis(seconds);

        return new TimeDataContainer(hours, minutes, seconds, millis);
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tvTime = findViewById(R.id.tvTime);
        tvPercent = findViewById(R.id.tvPercent);
        clockView = findViewById(R.id.clockView);
    }

    @Override
    protected void onResume() {
        super.onResume();

        timerHandler.postDelayed(timerRunnable, Delay);
    }

    @Override
    protected void onPause() {
        super.onPause();

        timerHandler.removeCallbacks(timerRunnable);
    }

    private void CalculateYearTime() {
        LocalDateTime now = LocalDateTime.now();
        LocalDateTime firstJanCurrentYear = LocalDateTime.of(now.getYear(), 1, 1, 0, 0);
        LocalDateTime firstJanNextYear = firstJanCurrentYear.minusYears(-1);

        long nsInYear = firstJanCurrentYear.until(firstJanNextYear, ChronoUnit.NANOS);
        double nsSinceYearStart = firstJanCurrentYear.until(now, ChronoUnit.NANOS);

        double percentOfYear = nsSinceYearStart / nsInYear;
        long msElapsed = (long) (percentOfYear * MsInDay);


        TimeDataContainer tdc = getDurationBreakdown(msElapsed);
        clockView.SetTime((int) tdc.hours, (int) tdc.minutes, (int) tdc.seconds);
        String timeString = String.format(Locale.ENGLISH, Format, tdc.hours, tdc.minutes, tdc.seconds, tdc.millis);
        tvTime.setText(timeString);

        tvPercent.setText(String.format(Locale.ENGLISH, "%.3f %%", percentOfYear * 100f));
    }
}

File: app/src/main/java/nl/jeroen/dayclock/MainActivity.kt
package nl.jeroen.dayclock

import android.os.Bundle
import android.os.Handler
import android.widget.TextView
import androidx.appcompat.app.AppCompatActivity
import java.time.LocalDateTime
import java.time.LocalTime
import java.time.format.DateTimeFormatter
import java.time.temporal.ChronoUnit
import java.util.*

class MainActivity : AppCompatActivity() {

    private var tvTime: TextView? = null
    private var tvTimeOfYear: TextView? = null
    private var tvPercent: TextView? = null
    private var clockView: ClockView? = null
    private var displayedYearSecond = -1
    private var timeFormatter: DateTimeFormatter? = null

    private val timerHandler = Handler()
    private val timerRunnable = object : Runnable {
        override fun run() {
            calculateYearTime()

            timerHandler.postDelayed(this, Period)
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        tvTime = findViewById(R.id.tvTime)
        tvTimeOfYear = findViewById(R.id.tvTimeOfYear)
        tvPercent = findViewById(R.id.tvPercent)
        clockView = findViewById(R.id.clockView)

        timeFormatter = DateTimeFormatter.ofPattern("HH:mm:ss.SSS")
    }

    override fun onResume() {
        super.onResume()

        timerHandler.postDelayed(timerRunnable, Delay)
    }

    override fun onPause() {
        super.onPause()

        timerHandler.removeCallbacks(timerRunnable)
    }

    private fun calculateYearTime() {
        val now = LocalDateTime.now()

        tvTime!!.text = now.format(timeFormatter)

        val firstJanCurrentYear = LocalDateTime.of(now.year, 1, 1, 0, 0)
        val firstJanNextYear = firstJanCurrentYear.plusYears(1)

        val nsInYear = firstJanCurrentYear.until(firstJanNextYear, ChronoUnit.NANOS)
        val nsSinceYearStart = firstJanCurrentYear.until(now, ChronoUnit.NANOS).toDouble()

        val percentOfYear = nsSinceYearStart / nsInYear
        val nsElapsed = (percentOfYear * NsInDay).toLong()

        val elapsedTime = LocalTime.ofNanoOfDay(nsElapsed)
        val currentSecond = elapsedTime.second

        if (displayedYearSecond != currentSecond) {
            displayedYearSecond = currentSecond

            clockView!!.setTime(elapsedTime.hour, elapsedTime.minute, currentSecond)
        }

        tvTimeOfYear!!.text = elapsedTime.format(timeFormatter)

        tvPercent!!.text = String.format(Locale.ENGLISH, "%.5f %%", percentOfYear * 100f)
    }

    companion object {

        private const val NsInDay = (24 * 60 * 60 * 1e9).toLong()
        private const val Delay = 0L
        private const val Period = 50L
    }
}
